### PR TITLE
Live Catch Module

### PR DESCRIPTION
**Pull request by MichaelMiner137**
When fishing in open water, fish will be flung at the player instead of the fish item.

### CODE REVIEW

File: gm4_live_catch/data/gm4_live_catch/functions/fish/cod/summon.mcfunction [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+
+# store nbt
+data modify storage gm4_live_catch:temp target set from entity @s
+
+# summon fish
+summon cod ~ ~1 ~ {Air:60,Health:1f,Tags:["gm4_lc_fish_new"]}
+
+# copy motion from item into fish
+data modify entity @e[type=cod,limit=1,sort=nearest,tag=gm4_lc_fish_new] Motion set from storage gm4_live_catch:temp target.Motion
```
Comment (Bloo-dev) [inline gm4_live_catch/data/gm4_live_catch/functions/fish/cod/summon.mcfunction:11]: You're not using this storage, so you might as well skip the storage entirely -- at least with the current setup. This might change in case you actually need to read more data from NBT.
```suggestion
data modify entity @e[type=cod,limit=1,sort=nearest,tag=gm4_lc_fish_new] Motion set from entity @s Motion
```

This also applies to the other types of fish.

File: gm4_live_catch/data/gm4_live_catch/functions/fish/cod/target.mcfunction [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# run from gm4_live_catch:fish/FISH_TYPE/catch
+
+
+# summon fish at item
+execute at @a[tag=gm4_lc_fisher] as @e[type=item,limit=1,sort=nearest,nbt={Age:0s}] at @s run function gm4_live_catch:fish/cod/summon
```
Comment (Bloo-dev) [inline gm4_live_catch/data/gm4_live_catch/functions/fish/cod/target.mcfunction:5]: You're not verifying that the item you're targeting is indeed a fish. In combination with the lack of a distance check this could lead to deletion of items by coincidence and is therefore unsafe.
You should verify that the item you're targeting is a fish and -- if possible -- is a fish that was just caught using a fishing rod. 

This also applies to the other types of fish.

File: gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/update.mcfunction [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# run from gm4_live_catch:fish/FISH_TYPE/summon
+
+
+# update
+data modify entity @s Air set value 60
+data modify entity @s Health set value 1f
```
Comment (Bloo-dev) [inline gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/update.mcfunction:6]: You have a storage set up, so writing these NBT values into storage would be more efficient, as the storage is merged onto the entity anyways.

**Review by Bloo-dev — CHANGES_REQUESTED** [review]
There are some changes that need to be made to the code to make this reliable and to ensure this is in line with the Gamemode 4 standard. I've left you an extensive list of comments and suggestions. Please note that a lot of the comments apply to all types of fish, but I've used `cod` as an example.

I have also tested this in-game and noticed that the motion on the fish is quite often insufficient to launch them onto land, resulting in the player having to chase their catch in the water. This should be adjusted by scaling up the motion vector by a sufficient amount.
Why does this only work in open water? Fish can also be caught when the bobber is not considered to be in open water; is there any lore reason I'm not aware of?

This module idea has received a lot of positive feedback in an internal call a couple of days ago.

File: gm4_live_catch/data/gm4_live_catch/functions/fish/cod/target.mcfunction (hunk 2 above)
Comment (BluePsychoRanger) [reply gm4_live_catch/data/gm4_live_catch/functions/fish/cod/target.mcfunction:5]: Additionally, you're not verifying if this player caught cod or a different fish, each fish should get a separate tag (as mentioned in an above comment) in the case that two players fish up two different fish at the same tick. Technically it wouldn't be visually detectable since the item is destroyed, so both players will get _a_ fish, but still.

Comment (BluePsychoRanger) [reply gm4_live_catch/data/gm4_live_catch/functions/fish/cod/target.mcfunction:5]: > You should verify that the item you're targeting is a fish and -- if possible -- is a fish that was just caught using a fishing rod.

It is not possible to detect if the fish was caught using a fishing rod. The code below will verify that it was cod that was caught and most likely to be from the fishing rod (not on the ground, count of 1, age of 0, and furthest instead of nearest as players are not standing right next to their fishing bobber), by a player who caught cod specifically.

```suggestion
execute at @a[tag=gm4_lc_fished_cod] as @e[type=item,limit=1,sort=furthest,nbt={Item:{id:"minecraft:cod",Count:1b},OnGround:0b,Age:0s}] at @s run function gm4_live_catch:fish/cod/summon
```

File: gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/summon.mcfunction [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# run from gm4_live_catch:fish/FISH_TYPE/target
+
+
+# store nbt
+data modify storage gm4_live_catch:temp target set from entity @s
```
Comment (BluePsychoRanger) [inline gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/summon.mcfunction:5]: This will allow you to merge the entire storage into the fish after adding the Air and Health modifications to the storage.

```suggestion
data modify storage gm4_live_catch:temp target.Motion set from entity @s Motion
```

File: gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/summon.mcfunction [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+# summon fish
+summon tropical_fish ~ ~1 ~ 
+execute as @e[type=tropical_fish,limit=1,sort=nearest] at @s run function gm4_live_catch:fish/tropical_fish/update
```
Comment (BluePsychoRanger) [inline gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/summon.mcfunction:9]: This code will ensure it's the correct fish. The position of `~ ~1 ~` is resolved in `tropical_fish/update`. Remember that at @s shouldn't be added unless you specifically need to check or modify it's position relative to itself
```suggestion
summon tropical_fish ~ ~ ~ 
execute as @e[type=tropical_fish,limit=1,distance=0] run function gm4_live_catch:fish/tropical_fish/update
```

File: gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/update.mcfunction (hunk 3 above)
Comment (BluePsychoRanger) [reply gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/update.mcfunction:6]:
```suggestion
data merge storage gm4_live_catch:temp {target:{Air:60,Health:1f}}
```

File: gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/update.mcfunction [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+tag @s add gm4_lc_fish_new
+
+# copy motion from item into fish
+data modify entity @s Motion set from storage gm4_live_catch:temp target.Motion
```
Comment (BluePsychoRanger) [inline gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/update.mcfunction:10]:
```suggestion
data modify entity @s {} merge from storage gm4_live_catch:temp target
tp @s ~ ~1 ~
```

File: gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/update.mcfunction [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# run from gm4_live_catch:fish/FISH_TYPE/summon
+
+
+# update
+data modify entity @s Air set value 60
+data modify entity @s Health set value 1f
+tag @s add gm4_lc_fish_new
```
Comment (BluePsychoRanger) [inline gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/update.mcfunction:7]: This tag is never used
```suggestion
```

File: gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/summon.mcfunction [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+# remove 
+tag @e[type=tropical_fish,tag=gm4_lc_fish_new] remove gm4_lc_fish_new
+
```
Comment (BluePsychoRanger) [inline gm4_live_catch/data/gm4_live_catch/functions/fish/tropical_fish/summon.mcfunction:13]: The tropical fish is never tagged
```suggestion
```

File: gm4_live_catch/data/gm4/advancements/live_catch.json [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+      "translate": "%1$s%3427655$s",
+      "with": [
+        "Watch Out!",
+        {"translate": "advancement.gm4.live_catch.title"}
+      ]
+    },
+    "description": {
+      "translate": "%1$s%3427655$s",
+      "with": [
+        "Catch a pufferfish and have it fly into your face.",
```
Comment (BluePsychoRanger) [inline gm4_live_catch/data/gm4/advancements/live_catch.json:17]: This description is too wordy in my opinion. It should say something like "Catch a pufferfish to the face"

**Review by BluePsychoRanger — CHANGES_REQUESTED** [review]
I like the idea of this module, but there needs to be some code optimizations. Similar to Bloo, I used the cod as an example to implement into the other fish types.

**Comment by MichaelMiner137** [comment]
Optimizations added.
Motion is now increased by 30% (thank you vdvman)

**Review by Bloo-dev — APPROVED** [review]

**Comment by EpyonProjects** [comment]
Reeling in fish does not seem to have motion.

**Comment by MichaelMiner137** [comment]
It will no longer affect modified fish.

**Review by Bloo-dev — CHANGES_REQUESTED** [review]
Tested this in game. Reeled in fish get a motion of 0, as stated by @EpyonProjects above.

This needs to be fixed before this can be merged.

**Comment by MichaelMiner137** [comment]
Motion now works properly when used for the first time.

**Comment by EpyonProjects** [comment]
Can confirm motion does work properly.  Seems a bit strong, but that's just my opinion.

**Comment by Dennis-0** [comment]
Did not find any issues when testing.

All functions contain comments with `FISH_TYPE` that should still be replaced
